src/server.rs

```diff
@@ -1978,3 +2058,108 @@ fn waits_until_zero() {
+            StringWriter {
+                buffer: String::new(),
+            }
+        }
+
+        fn get_output(self) -> String {
+            self.buffer
+        }
+    }
+
+    impl ServerStatsWriter for StringWriter {
+        fn write(&mut self, text: &str) {
+            self.buffer.push_str(&format!("{}\n", text));
+        }
+    }
+
+    #[test]
+    fn test_print_cache_hits_rate_default_server_stats() {
+        let stats = ServerStats::default();
+
+        let mut writer = StringWriter::new();
+        stats.print(&mut writer, false);
+
+        let output = writer.get_output();
+
+        assert!(output.contains("Cache hits rate                       -"));
+    }
+
+    #[test]
+    fn test_print_cache_hits_rate_server_stats() {
+        let mut cache_hits_counts = HashMap::new();
+        cache_hits_counts.insert("Rust".to_string(), 100);
+        cache_hits_counts.insert("C/C++".to_string(), 200);
+
+        let mut cache_misses_counts = HashMap::new();
+        cache_misses_counts.insert("Rust".to_string(), 50);
+        cache_misses_counts.insert("Cuda".to_string(), 300);
+
+        let stats = ServerStats {
+            cache_hits: PerLanguageCount {
+                counts: cache_hits_counts,
+                ..Default::default()
+            },
+            cache_misses: PerLanguageCount {
+                counts: cache_misses_counts,
+                ..Default::default()
+            },
+            ..Default::default()
+        };
+
+        let mut writer = StringWriter::new();
+        stats.print(&mut writer, false);
+
+        let output = writer.get_output();
+
+        assert!(output.contains("Cache hits rate                    46.15 %"));
+        assert!(output.contains("Cache hits rate (C/C++)           100.00 %"));
+        assert!(output.contains("Cache hits rate (Cuda)              0.00 %"));
+        assert!(output.contains("Cache hits rate (Rust)             66.67 %"));
+    }
+
+    #[test]
+    fn test_print_cache_hits_rate_advanced_server_stats() {
+        let mut cache_hits_counts = HashMap::new();
+        cache_hits_counts.insert("rust".to_string(), 50);
+        cache_hits_counts.insert("c/c++ [clang]".to_string(), 30);
+
+        let mut cache_misses_counts = HashMap::new();
+        cache_misses_counts.insert("rust".to_string(), 100);
+        cache_misses_counts.insert("cuda".to_string(), 70);
+
+        let stats = ServerStats {
+            cache_hits: PerLanguageCount {
+                adv_counts: cache_hits_counts,
+                ..Default::default()
+            },
+            cache_misses: PerLanguageCount {
+                adv_counts: cache_misses_counts,
+                ..Default::default()
+            },
+            ..Default::default()
+        };
+
+        let mut writer = StringWriter::new();
+        stats.print(&mut writer, true);
+
+        let output = writer.get_output();
+
+        assert!(output.contains("Cache hits rate                        -"));
```

Comment:
Interesting case where `PerLanguageCount::all()` uses `counts.values().sum()` and not `self.adv_counts.values().sum()` in case of printing advanced stats.

In practice should never happen because `PerLanguageCount::increment()` is used to set the counts.